src/OrleansSQLUtils/Storage/Provider/RelationalStorageProvider.cs

```diff
@@ -0,0 +1,550 @@
+        {
+            //If these are the same, it means no row was inserted or updated in the storage.
+            //Effectively it means the UPDATE or INSERT conditions failed due to ETag violation.
+            //Also if grainState.ETag storageVersion is null and storage comes back as null,
+            //it means two grains were activated an the other one succeeded in writing its state.
+            //
+            //NOTE: the storage could return also the new and old ETag (Version), but currently it doesn't.
+            if(storageVersion == grainVersion || storageVersion == string.Empty)
+            {
+                //TODO: Note that this error message should be canonical across back-ends.
+                return new InconsistentStateException($"Version conflict ({operation}): ServiceId={serviceId} ProviderName={providerName} GrainType={normalizedGrainType} GrainId={grainId} ETag={grainVersion}.");
+            }
+
+            return null;
+        }
+
+
+        /// <summary>
+        /// Writes a consistent log message from the given parameters.
+        /// </summary>
+        /// <param name="operationProlog">A free form prolog information to log.</param>
+        /// <param name="serviceId">Service Id.</param>
+        /// <param name="providerName">The name of this storage provider.</param>
+        /// <param name="version">The grain version.</param>
+        /// <param name="normalizedGrainType">Grain type without generics information.</param>
+        /// <param name="grainId">The grain ID.</param>
+        /// <param name="exceptionMessage">An optional exception message information to log.</param>
+        /// <returns>A log string to be printed.</returns>
+        private string LogString(string operationProlog, string serviceId, string providerName, string version, string normalizedGrainType, string grainId, string exceptionMessage = null)
+        {
+            const string Exception = " Exception=";
+            return $"{operationProlog}: ServiceId={serviceId} ProviderName={providerName} GrainType={normalizedGrainType} GrainId={grainId} ETag={version}{(exceptionMessage != null ? Exception + exceptionMessage : string.Empty)}.";
+        }
+
+
+        /// <summary>
+        /// Extracts the class base type without generic type information.
+        /// </summary>
+        /// <param name="classType">The class type information.</param>
+        /// <returns>The base type without generic type information.</returns>
+        /// <remarks>This is a work-around as Orleans gives grain type information in <see cref="ReadStateAsync(string, GrainReference, IGrainState)"/>
+        /// without generic type information and in <see cref="WriteStateAsync(string, GrainReference, IGrainState)"/> with it. This can be removed
+        /// when this is normalized.</remarks>
+        private static string ExtractClassBaseType(string classType)
```

Comment:
The current storage provider interface is very lacking (it has been discussed many times in the past). I'd rather we first fix it before resorting to workarounds.